### PR TITLE
HIVE-23215. Make FilterContext an interface.

### PR DESCRIPTION
**Pull request by omalley**
Here's a first pass at keeping a single copy of the selected data.

### CODE REVIEW

File: storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java [hunk 1]
```diff
@@ -359,4 +361,62 @@ public void ensureSize(int rows) {
       cols[i].ensureSize(rows, false);
     }
```
Comment (pgaref) [inline storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java:362]: Should we call updateSelected(rows) here as well?
According to getMaxSize() method, selected array should be able to hold the max number of rows.

File: storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java [hunk 2]
```diff
@@ -338,6 +339,7 @@ public void write(DataOutput arg0) throws IOException {
    *  - resets each column
    *  - inits each column
    */
+  @Override
   public void reset() {
```
Comment (pgaref) [inline storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java:343]: Can we also set the selected array to null here?

**Comment by pgaref** [comment]
Thanks for the patch @omalley ! LGTM overall left some comments

File: storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java (hunk 2 above)
Comment (omalley) [reply storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java:343]: My problem is that VectorizedRowBatch currently has an implicit invariant that selected is never null. Changing that assumption seems potentially problematic.

File: storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java (hunk 1 above)
Comment (omalley) [reply storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/VectorizedRowBatch.java:362]: Yeah, that is a good idea.

**Comment by omalley** [comment]
This seems to work well when I was working with the filters for ORC-577. A couple of comments:
* It feels a little worrisome letting the user pass in a selection array and they never know the right length without calling getSelectionSize first.
* We probably don't need the separate setSelectedInUser and setSelectedSize. All of the filters would just call setFilterContext to set both in a single call.

**Pull request closed without merging**